### PR TITLE
Add authenticated epoch

### PR DESCRIPTION
**Pull request by lxfind**
This PR adds authenticated epoch, and store it in the authority store.
To be filled

### CODE REVIEW

File: crates/sui-core/src/epoch/reconfiguration.rs [hunk 1]
```diff
@@ -52,7 +52,7 @@ where
             unreachable!();
         }
 
-        self.state.halted.store(true, Ordering::SeqCst);
+        self.state.halt_validator();
         while !self.state.batch_notifier.ticket_drained() {
```
Comment (gdanezis) [inline crates/sui-core/src/epoch/reconfiguration.rs:56]: Lets add a todo here to refactor to a condition variable pattern, this is a little fragile.

File: crates/sui-core/src/epoch/reconfiguration.rs [hunk 2]
```diff
@@ -52,7 +52,7 @@ where
             unreachable!();
         }
 
-        self.state.halted.store(true, Ordering::SeqCst);
+        self.state.halt_validator();
```
Comment (gdanezis) [inline crates/sui-core/src/epoch/reconfiguration.rs:55]: I think that when we halt the validator not only we need to wait for all tickets to be drained, but also we need to wait for all transactions to be in batches, and also in checkpoint proposals, right?

**Review by gdanezis — APPROVED** [review]
Good stuff!

File: crates/sui-types/src/messages.rs [hunk 3]
```diff
@@ -1770,3 +1771,75 @@ pub enum ExecuteTransactionResponse {
...
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct EpochInfo {
+    /// The epoch number of this epoch info.
+    /// Although we could derive it from the epoch number of `next_epoch_committee`, a potential
+    /// byzantine node may set next_epoch_committee.epoch as 0, which will make it very
+    /// inconvenient to obtain the epoch number.
+    pub epoch: EpochId,
+    pub next_epoch_committee: Committee,
+    // TODO: Add more fields.
```
[the remark this answers is not in the document]
Comment (lxfind) [reply crates/sui-types/src/messages.rs:1785]: Note that this data structure is created/signed at the end of an epoch.
But yes I agree that having start/end checkpoint is useful in this data structure.

**Pull request merged**